compiler/qsc_fir/src/ty.rs

```diff
@@ -0,0 +1,638 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use indenter::{indented, Format, Indented};
+use qsc_data_structures::span::Span;
+
+use crate::fir::{CallableKind, FieldPath, Functor, ItemId, Res};
+use std::{
+    collections::HashMap,
+    fmt::{self, Debug, Display, Formatter, Write},
+    rc::Rc,
+};
+
+fn set_indentation<'a, 'b>(
+    indent: Indented<'a, Formatter<'b>>,
+    level: usize,
+) -> Indented<'a, Formatter<'b>> {
+    indent.with_format(Format::Custom {
+        inserter: Box::new(move |_, f| {
+            for _ in 0..level {
+                write!(f, "    ")?;
+            }
+            Ok(())
+        }),
+    })
+}
+
+/// A type.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub enum Ty {
```

Comment:
Food for thought: the interpreter doesn't actually care about types and never reads them, so we could consider a later simplification of FIR that drops them.